packages/snap/src/helpers/credential-helpers.ts

```diff
@@ -1,47 +1,331 @@
+  let hasNextPage: boolean | undefined | null = false;
+
+  do {
+    const _variables: Record<string, unknown> = {
+      ...variables,
+      ...(endCursor && { after: endCursor }),
+    };
+
+    const { data, errors } = await client.execute<Res>(request, _variables);
+
+    if (errors) {
+      throw new TypeError(JSON.stringify(errors));
+    }
+
+    if (!data) {
+      throw new TypeError('No data returned from Ceramic');
+    }
+
+    encryptedVerifiableCredentials.push(data);
+
+    [endCursor, hasNextPage] = [
+      data?.verifiableCredentialIndex?.pageInfo?.endCursor,
+      data?.verifiableCredentialIndex?.pageInfo?.hasNextPage,
+    ];
+  } while (hasNextPage);
+
+  return encryptedVerifiableCredentials;
 };
 
+export const getAuthenticatedCeramicProvider = async (
+  pkHex?: string,
+  serverURL?: string,
+) => {
+  let privateKeyHex = pkHex ?? '';
+
+  if (!pkHex) {
+    const identityStorage = await getItemFromStore(StorageKeys.identity);
+
+    if (!identityStorage) {
+      throw new Error('Identity not created yet');
+    }
+
+    privateKeyHex = identityStorage.privateKeyHex;
+  }
+
+  const ceramicProvider = new CeramicProvider(privateKeyHex, serverURL);
+
+  await ceramicProvider.auth();
+
+  return ceramicProvider;
+};
+
+export class VCManager {
+  ceramicProvider: CeramicProvider;
+
+  constructor(ceramicProvider: CeramicProvider) {
+    this.ceramicProvider = ceramicProvider;
+  }
+
+  static async create(pkHex?: string, serverURL?: string) {
+    return new VCManager(
+      await getAuthenticatedCeramicProvider(pkHex, serverURL),
+    );
+  }
+
+  public async getDecryptedVCsByQueryHash(
+    queryHash: string,
+  ): Promise<W3CCredential[]> {
+    const client = this.ceramicProvider.client();
+
+    if (!client.did?.id) {
+      throw new TypeError('Client not authenticated');
+    }
+
+    const hashedClientDid = sha256(Buffer.from(client.did.id));
+
+    const hashedQueryHash = sha256(Buffer.from(queryHash));
+
+    const data = await loadAllCredentialsListPages<
+      GetVerifiableCredentialsByQueryHashQueryVariables,
+      GetVerifiableCredentialsByQueryHashQuery
+    >(
+      GetVerifiableCredentialsByQueryHash,
+      {
+        first: 1000,
+        queryHash: hashedQueryHash,
+        ownerDid: hashedClientDid,
+      },
+      this.ceramicProvider,
+    );
+
+    const executionResult = data
+      .map((el) => el.verifiableCredentialIndex?.edges)
+      .flat();
+
+    return await Promise.all(
+      executionResult.map(async (el) => {
+        const encryptedVC = el?.node?.data as string;
+
+        return await this.ceramicProvider.decrypt<W3CCredential>(encryptedVC);
+      }),
+    );
+  }
+
+  public async getDecryptedVCsByOffer(
+    offer: ClaimOffer,
+  ): Promise<W3CCredential[]> {
+    const claimIds = getClaimIdsFromOffer(offer);
+
+    const encryptedVCs = await Promise.all(
+      claimIds.map(async (claimId) => {
+        const hashedClaimId = sha256(Buffer.from(claimId));
```

Comment:
You should add some account-specific entropy to the input, otherwise all VCs of the same type will have the same hashedClaimId

Reply:
yeah, but user will get only his VCs, cause it's filtered by ownerId. Is it ok? Or should we keep every VC entry unique?

Reply:
It's ok. The additional entropy will prevent someone from counting the total number of credentials issued by some particular issuer, e.g. members of the same org